📖 document metrics scraping and enable metrics services via annotations (#4247)
kubernetes-sigs/cluster-api: docs/book/src/reference/metrics.md

## Metrics

> By default, controller-runtime builds a global prometheus registry and

Reviewer: ```suggestion
By default, controller-runtime builds a global Prometheus registry and
```

publishes a collection of performance metrics for each controller.

### Protecting the Metrics

These metrics are protected by [kube-auth-proxy](https://github.com/brancz/kube-rbac-proxy)
by default.

You will need to grant permissions to your Prometheus server so that it can

> scrape the protected metrics. To achieve that, you can create a `clusterRole` and a 
> `clusterRoleBinding` to bind to the service account that your Prometheus server uses.

Reviewer 2: nits:
```suggestion
scrape the protected metrics. To achieve that, you can create a `ClusterRole` and a 
`ClusterRoleBinding` to bind to the service account that your Prometheus server uses.
```
following [Use upper camel case for API objects](https://kubernetes.io/docs/contribute/style/style-guide/#use-upper-camel-case-for-api-objects)

> Create a YAML file named `capi-metrics-reader-clusterrole.yaml` with following content

Reviewer 2: ```suggestion
Create a manifest named `capi-metrics-reader-clusterrole.yaml` with the following content:
```

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: capi-metrics-reader
rules:
- nonResourceURLs: ["/metrics"]

>   verbs: ["get"]

Reviewer 3: We may as well add this to the infra components yaml and remove this step for the end user.

Probably makes sense to leave out the cluster role binding as we don't know the namespace Prometheus may be deployed into.

Author: That's the reason i documented it this way

Reviewer: @randomvariable so did I get it right that you suggest removing this YAML here and adding it to the infra component YAMLs of all providers? Maybe I'm missing something but in case we really want to add this to our deployments, it looks to me like we need this ClusterRole only once. 

But I'm really not sure if we should add this to our YAMLs so that it's deployed everywhere. Imho Prometheus and RBAC setups can vary and (as far as I'm aware) there was no recurring demand for this in Slack. I assume nobody is really missing this in our YAMLs at the moment.

I have no strong opinion against adding the ClusteRole to our YAMLs, but if we do we should do it right and adding it to all infra providers seems redundant to me.

Author: if #4640 will be merged, there is no need to add these objects to the infra components yaml. Let's wait what happens to #4640 and discuss again

```

> and apply the `clusterRole` with

Reviewer 2: nit:
```suggestion
and apply the `ClusterRole` with
```

```bash
kubectl apply -f capi-metrics-reader-clusterrole.yaml
```

> You can run the following kubectl command to create a `clusterRoleBinding` and grant access on the `/metrics` endpoint to your Prometheus instance (`<namespace>` must be the namespace where your Prometheus instance is running. `<service-account-name>` must be the service account name which is configured in your Prometheus instance).

Reviewer 2: nit
```suggestion
You can run the following kubectl command to create a `ClusterRoleBinding` and grant access on the `/metrics` endpoint to your Prometheus instance (`<namespace>` must be the namespace where your Prometheus instance is running. `<service-account-name>` must be the service account name which is configured in your Prometheus instance).
```

```bash
kubectl create clusterrolebinding capi-metrics-reader --clusterrole=capi-metrics-reader --serviceaccount=<namespace>:<service-account-name>
```

### Scraping the Metrics with Prometheus

> To scrape metrics, your Prometheus instance need at least following [`kubernetes_sd_config`](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#kubernetes_sd_config) section.

Reviewer: ```suggestion
To scrape metrics, your Prometheus instance needs at least the following [`kubernetes_sd_config`](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#kubernetes_sd_config) section.
```

          # if service endpoints use their own CA (e.g. via cert-manager) which aren't
          # signed by the cluster-internal CA we must skip the cert validation
          insecure_skip_verify: true
        bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
        kubernetes_sd_configs:
          - role: endpoints
        relabel_configs:
          - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_scrape]
            action: keep
            regex: true
          - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_scheme]
            action: replace
            target_label: __scheme__
            regex: (https?)
          - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_path]
            action: replace
            target_label: __metrics_path__
            regex: (.+)
          - source_labels: [__address__, __meta_kubernetes_service_annotation_prometheus_io_port]
            action: replace
            target_label: __address__
            regex: ([^:]+)(?::\d+)?;(\d+)
            replacement: $1:$2
          - action: labelmap
            regex: __meta_kubernetes_service_label_(.+)
```

> You are no able to check for metrics in your Prometheus instance. To verify, you could search with e.g. `{namespace="capi-system"}` to get all metrics from components running in `capi-system` Namespace.

Reviewer 2: nits:
```suggestion
You are no able to check for metrics in your Prometheus instance. To verify, you could search with e.g. `{namespace="capi-system"}` to get all metrics from components running in the `capi-system` namespace.
```

(2<sup>nd</sup> tweak because SIG Docs would typically write “namespace” in this context all-lowercase (and if localizing we'd pick the local term rather than referring to `Namespace`)

> You are no able to check for metrics in your Prometheus instance. To verify, you could search with e.g. `{namespace="capi-system"}` to get all metrics from components running in the `capi-system` namespace.

Reviewer: ```suggestion
You are now able to check for metrics in your Prometheus instance. To verify, you could search with e.g. `{namespace="capi-system"}` to get all metrics from components running in the `capi-system` namespace.
```
